### ISSUE TITLE
Loading React more than once is causing context issues w/ react-router.  

### ISSUE THREAD

**Issue by fricke** [milankinen/livereactload#8]
Hope you can help or point me in the right direction, but I've been battling w/ this most of the day and finally realized it wasn't anything in my code.  I took your 05 example and swapped out it with any example from here:

https://github.com/rackt/react-router/tree/master/examples

And it works fine on first render, but if you make a change I get a pretty horrendous message.  Very much to do w/ context and the new way of handling things in React 0.13.  

```
warning: owner-based and parent-based contexts differ (values: `2` vs `1`) for key (routeDepth) while mounting RouteHandler (see: http://fb.me/react-context-by-parent)
:3000/static/bundle.js:23498 Warning: Failed Context Types: Required context `router` was not specified in `Link`. Check the render method of `App`.
:3000/static/bundle.js:23498 Warning: owner-based and parent-based contexts differ (values: `undefined` vs `function (props, context) {
      // This constructor is overridden by mocks. The argument is used
      // by mocks to assert on what gets mounted.

      if ("production" !== process.env.NODE_ENV) {
        ("production" !== process.env.NODE_ENV ? warning(
          this instanceof Constructor,
          'Something is calling a React component directly. Use a factory or ' +
          'JSX instead. See: http://fb.me/react-legacyfactory'
        ) : null);
      }

      // Wire up auto-binding
      if (this.__reactAutoBindMap) {
        bindAutoBindMethods(this);
      }

      this.props = props;
      this.context = context;
      this.state = null;

      // ReactClasses doesn't have constructors. Instead, they use the
      // getInitialState and componentWillMount methods for initialization.

      var initialState = this.getInitialState ? this.getInitialState() : null;
      if ("production" !== process.env.NODE_ENV) {
        // We allow auto-mocks to proceed as if they're returning null.
        if (typeof initialState === 'undefined' &&
            this.getInitialState._isMockFunction) {
          // This is probably bad practice. Consider warning here and
          // deprecating this convenience.
          initialState = null;
        }
      }
      ("production" !== process.env.NODE_ENV ? invariant(
        typeof initialState === 'object' && !Array.isArray(initialState),
        '%s.getInitialState(): must return an object or null',
        Constructor.displayName || 'ReactCompositeComponent'
      ) : invariant(typeof initialState === 'object' && !Array.isArray(initialState)));

      this.state = initialState;
    }`) for key (router) while mounting Link (see: http://fb.me/react-context-by-parent)
:3000/static/bundle.js:1896 Uncaught TypeError: Cannot read property 'makeHref' of undefined
:3000/static/bundle.js:23498 Warning: Failed Context Types: Required context `router` was not specified in `Link`. Check the render method of `App`.
:3000/static/bundle.js:23498 Warning: owner-based and parent-based contexts differ (values: `undefined` vs `function (props, context) {
      // This constructor is overridden by mocks. The argument is used
      // by mocks to assert on what gets mounted.

      if ("production" !== process.env.NODE_ENV) {
        ("production" !== process.env.NODE_ENV ? warning(
          this instanceof Constructor,
          'Something is calling a React component directly. Use a factory or ' +
          'JSX instead. See: http://fb.me/react-legacyfactory'
        ) : null);
      }

      // Wire up auto-binding
      if (this.__reactAutoBindMap) {
        bindAutoBindMethods(this);
      }

      this.props = props;
      this.context = context;
      this.state = null;

      // ReactClasses doesn't have constructors. Instead, they use the
      // getInitialState and componentWillMount methods for initialization.

      var initialState = this.getInitialState ? this.getInitialState() : null;
      if ("production" !== process.env.NODE_ENV) {
        // We allow auto-mocks to proceed as if they're returning null.
        if (typeof initialState === 'undefined' &&
            this.getInitialState._isMockFunction) {
          // This is probably bad practice. Consider warning here and
          // deprecating this convenience.
          initialState = null;
        }
      }
      ("production" !== process.env.NODE_ENV ? invariant(
        typeof initialState === 'object' && !Array.isArray(initialState),
        '%s.getInitialState(): must return an object or null',
        Constructor.displayName || 'ReactCompositeComponent'
      ) : invariant(typeof initialState === 'object' && !Array.isArray(initialState)));

      this.state = initialState;
    }`) for key (router) while mounting Link (see: http://fb.me/react-context-by-parent)
:3000/static/bundle.js:10735 Uncaught TypeError: Cannot read property '_currentElement' of null
```

Basically what's happening is that on the second round I'm losing my context and it's blowing up react-router as it expects that.  I digged further and realized that bundle.js is added again to the page on each livereload notify which may be the issue.  I just came off of webpack dev b/c of some issues I had, but it seems like what is done there is very specific module loading and it seems like that is what is going on here, but again context is somehow not coming into play on livereactload refresh.  However, maybe I'm missing something obvious?  

Any help would be greatly appreciated!  Thanks so much

Chris

**Comment by milankinen** [comment]
React Router seems to be using React as a peerDependency so this might have a same root cause as issue https://github.com/milankinen/livereactload/issues/7. My guess is that there might be also other things with react-router as well, but we'll see that after I get that peer dependency issue fixed.

**Comment by milankinen** [comment]
I've published beta version `0.2.3` which should solve the problem. If you can handle the fact that you need to manually refresh the page if your routes are changed then this kind of implementation seems to work with me: https://gist.github.com/milankinen/e4ebeb381b8ce59c4291 

(You must manually set your `livereactload` package version to 0.2.3 and set livereactload as a global transform)

**Comment by milankinen** [comment]
I have had no issues since `0.2.4` release so I close this issue

**Comment by davertron** [comment]
I am still seeing this issue using livereactload 0.5.1 and react-router 0.13.3. If I try to use livereactload as a global transform I don't get the crazy message above, but I do still see the following message:

Uncaught Error: Invariant Violation: Router is already running

When this happens the routes stop working completely. Any ideas? I'm using livereactload with browserify and gulp, if that helps.

**Comment by defly** [comment]
Hi, I am seeing this issue too.("react-router": "^0.13.3", "livereactload": "^0.5.2", "react": "^0.13.3") 

Here the code:

```javascript
var routes = (
  <Route handler={App} path="/">
    <DefaultRoute handler={Home} />
    <Route name="users" handler={Users} />
    <NotFoundRoute handler={NotFound}/>
  </Route>
);

Router.run(routes, Router.HistoryLocation, function(Handler) {
  React.render(<Handler/>, mountNode);
});

```
```javascript
import React from 'react';
import {RouteHandler} from 'react-router';

class App extends React.Component {
  render() {
    return (
        <div>
          <h1>App1</h1>
          <RouteHandler/>
        </div>
    );
  }
}

export default App;

```

After changing the code in DevTools I see:

```
Warning: Failed Context Types: Required context `routeDepth` was not specified in `RouteHandler`. Check the render method of `App`.
warning.js:48 Warning: Failed Context Types: Required context `router` was not specified in `RouteHandler`. Check the render method of `App`.
warning.js:48 Warning: owner-based and parent-based contexts differ (values: `undefined` vs `1`) for key (routeDepth) while mounting RouteHandler (see: http://fb.me/react-context-by-parent)
warning.js:48 Warning: owner-based and parent-based contexts differ (values: `undefined` vs `function (props, context) {
      // This constructor is overridden by mocks. The argument is used
      // by mocks to assert on what gets mounted.

      if ("production" !== process.env.NODE_ENV) {
        ("production" !== process.env.NODE_ENV ? warning(
          this instanceof Constructor,
          'Something is calling a React component directly. Use a factory or ' +
          'JSX instead. See: https://fb.me/react-legacyfactory'
        ) : null);
      }

      // Wire up auto-binding
      if (this.__reactAutoBindMap) {
        bindAutoBindMethods(this);
      }

      this.props = props;
      this.context = context;
      this.state = null;

      // ReactClasses doesn't have constructors. Instead, they use the
      // getInitialState and componentWillMount methods for initialization.

      var initialState = this.getInitialState ? this.getInitialState() : null;
      if ("production" !== process.env.NODE_ENV) {
        // We allow auto-mocks to proceed as if they're returning null.
        if (typeof initialState === 'undefined' &&
            this.getInitialState._isMockFunction) {
          // This is probably bad practice. Consider warning here and
          // deprecating this convenience.
          initialState = null;
        }
      }
      ("production" !== process.env.NODE_ENV ? invariant(
        typeof initialState === 'object' && !Array.isArray(initialState),
        '%s.getInitialState(): must return an object or null',
        Constructor.displayName || 'ReactCompositeComponent'
      ) : invariant(typeof initialState === 'object' && !Array.isArray(initialState)));

      this.state = initialState;
    }`) for key (router) while mounting RouteHandler (see: http://fb.me/react-context-by-parent)
RouteHandler.js:67Uncaught TypeError: Cannot read property 'getRouteAtDepth' of undefinedcreateChildRouteHandler @ RouteHandler.js:67render @ RouteHandler.js:82ReactCompositeComponentMixin._renderValidatedComponentWithoutOwnerOrContext @ ReactCompositeComponent.js:789ReactCompositeComponentMixin._renderValidatedComponent @ ReactCompositeComponent.js:816ReactPerf.measure.wrapper @ ReactPerf.js:70ReactCompositeComponentMixin.mountComponent @ ReactCompositeComponent.js:237ReactPerf.measure.wrapper @ ReactPerf.js:70ReactReconciler.mountComponent @ ReactReconciler.js:38__REACT_INSPECTOR_RUNTIME__0_13_1.React.ReactReconciler.mountComponent @ VM16021:106ReactMultiChild.Mixin.mountChildren @ ReactMultiChild.js:192ReactDOMComponent.Mixin._createContentMarkup @ ReactDOMComponent.js:289ReactDOMComponent.Mixin.mountComponent @ ReactDOMComponent.js:199ReactReconciler.mountComponent @ ReactReconciler.js:38__REACT_INSPECTOR_RUNTIME__0_13_1.React.ReactReconciler.mountComponent @ VM16021:106ReactCompositeComponentMixin.mountComponent @ ReactCompositeComponent.js:247ReactPerf.measure.wrapper @ ReactPerf.js:70ReactReconciler.mountComponent @ ReactReconciler.js:38__REACT_INSPECTOR_RUNTIME__0_13_1.React.ReactReconciler.mountComponent @ VM16021:106ReactCompositeComponentMixin.mountComponent @ ReactCompositeComponent.js:247ReactPerf.measure.wrapper @ ReactPerf.js:70ReactReconciler.mountComponent @ ReactReconciler.js:38__REACT_INSPECTOR_RUNTIME__0_13_1.React.ReactReconciler.mountComponent @ VM16021:106ReactCompositeComponentMixin.mountComponent @ ReactCompositeComponent.js:247ReactPerf.measure.wrapper @ ReactPerf.js:70ReactReconciler.mountComponent @ ReactReconciler.js:38__REACT_INSPECTOR_RUNTIME__0_13_1.React.ReactReconciler.mountComponent @ VM16021:106mountComponentIntoNode @ ReactMount.js:248Mixin.perform @ Transaction.js:134batchedMountComponentIntoNode @ ReactMount.js:269Mixin.perform @ Transaction.js:134ReactDefaultBatchingStrategy.batchedUpdates @ ReactDefaultBatchingStrategy.js:66batchedUpdates @ ReactUpdates.js:110ReactMount._renderNewRootComponent @ ReactMount.js:404ReactPerf.measure.wrapper @ ReactPerf.js:70ReactMount.render @ ReactMount.js:493ReactPerf.measure.wrapper @ ReactPerf.js:70(anonymous function) @ entryClient.jsx:23React.createClass.statics.run.dispatchHandler @ createRouter.js:381(anonymous function) @ createRouter.js:349(anonymous function) @ Transition.js:69(anonymous function) @ Transition.js:69Transition.to @ Transition.js:72(anonymous function) @ createRouter.js:348Transition.from @ Transition.js:51dispatch @ createRouter.js:345refresh @ createRouter.js:396run @ createRouter.js:392runRouter @ runRouter.js:45./web/app/entryClient.jsx../App.jsx @ entryClient.jsx:22s @ _prelude.js:1e @ _prelude.js:1(anonymous function) @ _prelude.js:1
warning.js:48 Warning: Failed Context Types: Required context `routeDepth` was not specified in `RouteHandler`. Check the render method of `App`.
warning.js:48 Warning: Failed Context Types: Required context `router` was not specified in `RouteHandler`. Check the render method of `App`.
warning.js:48 Warning: owner-based and parent-based contexts differ (values: `undefined` vs `1`) for key (routeDepth) while mounting RouteHandler (see: http://fb.me/react-context-by-parent)
warning.js:48 Warning: owner-based and parent-based contexts differ (values: `undefined` vs `function (props, context) {
      // This constructor is overridden by mocks. The argument is used
      // by mocks to assert on what gets mounted.

      if ("production" !== process.env.NODE_ENV) {
        ("production" !== process.env.NODE_ENV ? warning(
          this instanceof Constructor,
          'Something is calling a React component directly. Use a factory or ' +
          'JSX instead. See: https://fb.me/react-legacyfactory'
        ) : null);
      }

      // Wire up auto-binding
      if (this.__reactAutoBindMap) {
        bindAutoBindMethods(this);
      }

      this.props = props;
      this.context = context;
      this.state = null;

      // ReactClasses doesn't have constructors. Instead, they use the
      // getInitialState and componentWillMount methods for initialization.

      var initialState = this.getInitialState ? this.getInitialState() : null;
      if ("production" !== process.env.NODE_ENV) {
        // We allow auto-mocks to proceed as if they're returning null.
        if (typeof initialState === 'undefined' &&
            this.getInitialState._isMockFunction) {
          // This is probably bad practice. Consider warning here and
          // deprecating this convenience.
          initialState = null;
        }
      }
      ("production" !== process.env.NODE_ENV ? invariant(
        typeof initialState === 'object' && !Array.isArray(initialState),
        '%s.getInitialState(): must return an object or null',
        Constructor.displayName || 'ReactCompositeComponent'
      ) : invariant(typeof initialState === 'object' && !Array.isArray(initialState)));

      this.state = initialState;
    }`) for key (router) while mounting RouteHandler (see: http://fb.me/react-context-by-parent)
ReactCompositeComponent.js:744 Uncaught TypeError: Cannot read property '_currentElement' of null
```

Thank you!

**Comment by omaksi** [comment]
I am seeing the issue as well 
("react-router": "^0.13.3", "livereactload": "^0.5.2", "react": "^0.13.3")

Anyone know if this is an issue with livereactload or react-router?

**Comment by remaininlight** [comment]
I have never used livereactreload and I'm getting this message. Arrived here from google and thought you guys should know..

**Comment by milankinen** [comment]
Do you call `Router.run` inside the `window.onload` callback? 

LiveReactload loads the whole bundle and executes all the JavaScript during the reload. If you are not calling `Router.run` inside the `window.onload` it'll be executed multiple times, thus probably causing that error message.

Please see this example: https://gist.github.com/milankinen/b37314362e82314c4da3
